File: frontend/src/api/analysis.ts
// =====================================================
// src/api/analysis.ts
// 数据分析模块 API 接口层
// 严格遵循 API 接口规范文档 v1.0 第七章
// 包含: RFM分析、K-Means聚类、重新计算、关联规则、用户画像
// 当前使用 Mock 数据，后端就绪后设置 USE_MOCK = false
// =====================================================

import request from '@/utils/request'
import type {
  ApiResponse,
  RFMResult,
  ClusterResult,
  AssociationRule,
  UserProfile,
} from '@/types/api'

// ==================== Mock 开关 ====================
const USE_MOCK = false

// ==================== Mock 数据 ====================

/** Mock RFM 分析结果 - 对应 API 7.1 */
const mockRFMData = (): RFMResult => {
  const segments = ['高价值用户', '忠诚用户', '潜力用户', '新用户', '流失用户']
  const scores = Array.from({ length: 50 }, (_, i) => ({
    user_id: i + 1,
    recency: Math.floor(Math.random() * 5) + 1,
    frequency: Math.floor(Math.random() * 5) + 1,
    monetary: Math.floor(Math.random() * 5) + 1,
    segment: segments[Math.floor(Math.random() * segments.length)],
  }))

  return {
    distribution: {
      labels: segments,
      values: [158, 325, 287, 412, 198],
    },
    scores,
  }
}

/** Mock K-Means 聚类结果 - 对应 API 7.2 */
const mockClusterData = (): ClusterResult => {
  const clusterNames = [
    '高消费活跃用户',
    '中等消费用户',
    '低频低消费用户',
    '新用户群体',
    '流失预警用户',
  ]

  return {
    clusters: clusterNames.map((name, label) => ({
      label,
      name,
      count: [150, 320, 280, 400, 200][label],
      center: [
        [2.5, 8.3, 1250.0],
        [5.0, 4.2, 580.0],
        [8.2, 1.5, 120.0],
        [3.0, 2.0, 200.0],
        [9.0, 0.8, 50.0],
      ][label],
      points: Array.from({ length: 20 }, (_, i) => ({
        user_id: label * 20 + i + 1,
        x: parseFloat((Math.random() * 3 + 1).toFixed(2)),
        y: parseFloat((Math.random() * 8 + 0.5).toFixed(2)),
        z: parseFloat((Math.random() * 1500 + 50).toFixed(2)),
      })),
    })),
  }
}

/** Mock 关联规则 - 对应 API 7.4（商品相关，预留） */
const mockAssociationRules = (): { rules: AssociationRule[] } => {
  const products = [
    { id: 'P1001', name: '无线蓝牙耳机' },
    { id: 'P1002', name: '充电宝' },
    { id: 'P1003', name: '手机壳' },
    { id: 'P1004', name: 'Type-C数据线' },
    { id: 'P1005', name: '智能手表' },
    { id: 'P1006', name: '无线鼠标' },
    { id: 'P1007', name: '机械键盘' },
    { id: 'P1008', name: '平板支架' },
  ]

  const rules: AssociationRule[] = [
    {
      antecedents: ['P1001'],
      consequents: ['P1002'],
      antecedent_names: ['无线蓝牙耳机'],
      consequent_names: ['充电宝'],
      support: 0.052,
      confidence: 0.65,
      lift: 2.3,
    },
    {
      antecedents: ['P1001'],
      consequents: ['P1004'],
      antecedent_names: ['无线蓝牙耳机'],
      consequent_names: ['Type-C数据线'],
      support: 0.048,
      confidence: 0.58,
      lift: 1.9,
    },
    {
      antecedents: ['P1005'],
      consequents: ['P1007'],
      antecedent_names: ['智能手表'],
      consequent_names: ['机械键盘'],
      support: 0.035,
      confidence: 0.52,
      lift: 2.1,
    },
    {
      antecedents: ['P1006', 'P1007'],
      consequents: ['P1008'],
      antecedent_names: ['无线鼠标', '机械键盘'],
      consequent_names: ['平板支架'],
      support: 0.028,
      confidence: 0.61,
      lift: 2.6,
    },
  ]

  return { rules }
}

/** Mock 用户画像 - 对应 API 7.5 */
const mockUserProfile = (userId: number): UserProfile => ({
  user_id: userId,
  nickname: ['张三', '李四', '王五', '赵六'][userId % 4],
  tags: ['高价值用户', '电子产品爱好者', '周末购物', '促销敏感型'],
  prefer_categories: [
    { name: '电子产品', count: 15 },
    { name: '服装', count: 8 },
    { name: '家居用品', count: 5 },
    { name: '食品饮料', count: 3 },
    { name: '运动户外', count: 2 },
  ],
  active_hours: [10, 11, 14, 20, 21, 22],
  avg_order_value: parseFloat((Math.random() * 300 + 50).toFixed(2)),
  total_orders: Math.floor(Math.random() * 30) + 1,
  days_since_last_purchase: Math.floor(Math.random() * 30) + 1,
})

/** 模拟网络延迟 */
const mockDelay = <T>(data: T, delay = 400): Promise<T> =>
  new Promise((resolve) => setTimeout(() => resolve(data), delay))

// ==================== API 接口函数 ====================

/**
 * 7.1 RFM 用户价值分析
 * 接口: GET /analysis/rfm
 * 描述: 获取 RFM 分析结果（用户分布 + 分数列表）
 *
 * RFM 模型说明:
 *   R(Recency)  - 最近购买时间，值越小越好（1-5分）
 *   F(Frequency) - 购买频率，值越大越好（1-5分）
 *   M(Monetary)  - 消费金额，值越大越好（1-5分）
 *
 * @returns Promise<ApiResponse<RFMResult>>
 */
export const getRFMAnalysis = (): Promise<ApiResponse<RFMResult>> => {
  if (USE_MOCK) {
    return mockDelay({
      code: 200,
      message: 'success',
      data: mockRFMData(),
    })
  }
  return request.get('/analysis/rfm')
}

/**
 * 7.2 K-Means 用户分群
 * 接口: GET /analysis/cluster
 * 描述: 获取 K-Means 聚类结果（用于散点图/雷达图）
 *
 * 聚类说明:
 *   使用 K-Means 算法对用户进行分群（默认 k=5）
 *   基于 RFM 三维特征进行聚类
 *
 * @returns Promise<ApiResponse<ClusterResult>>
 */
export const getClusterAnalysis = (): Promise<ApiResponse<ClusterResult>> => {
  if (USE_MOCK) {
    return mockDelay({
      code: 200,
      message: 'success',
      data: mockClusterData(),
    }, 500)
  }
  return request.get('/analysis/cluster')
}

/**
 * 7.3 触发分析重新计算
 * 接口: POST /analysis/recalculate
 * 描述: 触发后台重新计算 RFM 和聚类结果（耗时操作）
 *
 * @param type - 计算类型: 'rfm' | 'cluster' | 'all'
 * @param clusters - 聚类数量（仅 type='cluster' 时有效）
 * @returns Promise<ApiResponse<null>>
 */
export const recalculateAnalysis = (
  type: 'rfm' | 'cluster' | 'all' = 'all',
  clusters = 5
): Promise<ApiResponse<null>> => {
  if (USE_MOCK) {
    return mockDelay(
      {
        code: 200,
        message: `${type === 'all' ? 'RFM分析和聚类' : type === 'rfm' ? 'RFM分析' : '聚类分析'}重新计算完成`,
        data: null,
      },
      1500
    )
  }
  return request.post('/analysis/recalculate', { type, clusters })
}

/**
 * 7.4 关联规则挖掘（Apriori 算法）
 * 接口: GET /analysis/association
 * 描述: 获取商品关联规则（用于推荐模块）
 *
 * @param minSupport - 最小支持度，默认 0.01
 * @param minConfidence - 最小置信度，默认 0.5
 * @param limit - 返回数量，默认 20
 * @returns Promise<ApiResponse<{rules: AssociationRule[]}>>
 */
export const getAssociationRules = (
  minSupport = 0.01,
  minConfidence = 0.5,
  limit = 20
): Promise<ApiResponse<{ rules: AssociationRule[] }>> => {
  if (USE_MOCK) {
    const data = mockAssociationRules()
    return mockDelay({
      code: 200,
      message: 'success',
      data: { rules: data.rules.slice(0, limit) },
    })
  }
  return request.get('/analysis/association', {
    params: { min_support: minSupport, min_confidence: minConfidence, limit },
  })
}

/**
 * 7.5 用户画像
 * 接口: GET /analysis/profile/{user_id}
 * 描述: 获取指定用户的画像数据（标签、偏好分类、活跃时段等）
 *
 * @param userId - 用户ID
 * @returns Promise<ApiResponse<UserProfile>>
 */
export const getUserProfile = (userId: number): Promise<ApiResponse<UserProfile>> => {
  if (USE_MOCK) {
    return mockDelay({
      code: 200,
      message: 'success',
      data: mockUserProfile(userId),
    })
  }
  return request.get(`/analysis/profile/${userId}`)
}

File: backend/app/api/analysis.py
# =============================================================================
# 智能零售用户行为分析系统 - 数据分析模块 API 路由
# =============================================================================
# 5个接口:
#   7.1 GET  /analysis/rfm          - RFM用户价值分析
#   7.2 GET  /analysis/cluster      - K-Means用户分群
#   7.3 POST /analysis/recalculate  - 触发重新计算
#   7.4 GET  /analysis/association  - Apriori关联规则
#   7.5 GET  /analysis/profile/{id} - 用户画像
# =============================================================================

from fastapi import APIRouter, Depends, Query, HTTPException, status
from sqlalchemy.orm import Session
from typing import Dict, Optional

from app.schemas.analysis import (
    RecalculateRequest, RFMResult, RFMDistribution, RFMScoreItem,
    ClusterResultResponse, ClusterItem, ClusterPoint,
    AssociationResult, AssociationRule,
    UserProfileResponse, PreferCategory
)
from app.models.rfm_score import RfmScore
from app.models.cluster_result import ClusterResult
from app.models.base import get_db
from app.utils.response_utils import success_response, error_response
from app.utils.jwt_utils import get_current_user
from app.services.analysis_service import calculate_rfm, calculate_cluster, calculate_association, get_user_profile


router = APIRouter(prefix="/analysis", tags=["数据分析"])


# ======================== 7.1 RFM 用户价值分析 ========================

@router.get("/rfm", summary="RFM用户价值分析")
async def get_rfm(
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """
    获取RFM分析结果
    
    如果数据库中没有RFM数据，自动触发计算并存储
    """
    scores = db.query(RfmScore).all()
    if not scores:
        # 没有数据，自动计算
        result = calculate_rfm(db)
    else:
        # 已有数据，直接读取
        segments_count = {}
        score_list = []
        for s in scores:
            seg = s.segment or "流失用户"
            segments_count[seg] = segments_count.get(seg, 0) + 1
            score_list.append(RFMScoreItem(
                user_id=s.user_id, recency=s.recency or 1, frequency=s.frequency or 1,
                monetary=s.monetary or 1, segment=seg
            ).model_dump())
        labels = ["高价值用户", "忠诚用户", "潜力用户", "流失用户"]
        result = {
            "distribution": {"labels": labels, "values": [segments_count.get(l, 0) for l in labels]},
            "scores": score_list
        }
    
    return success_response(data=RFMResult(
        distribution=RFMDistribution(**result["distribution"]),
        scores=[RFMScoreItem(**s) for s in result["scores"]]
    ).model_dump())


# ======================== 7.2 K-Means 用户分群 ========================

@router.get("/cluster", summary="K-Means用户分群")
async def get_cluster(
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """获取K-Means聚类结果（散点图/雷达图数据）"""
    records = db.query(ClusterResult).all()
    if not records:
        # 确保RFM存在
        if not db.query(RfmScore).count():
            calculate_rfm(db)
        result = calculate_cluster(db)
        records = db.query(ClusterResult).all()
    else:
        # 构建聚类结果
        from collections import defaultdict
        rfm_map = {s.user_id: s for s in db.query(RfmScore).all()}
        groups = defaultdict(list)
        for c in records:
            rfm = rfm_map.get(c.user_id)
            groups[c.cluster_label].append({
                "user_id": c.user_id,
                "x": float(rfm.recency) if rfm else 0,
                "y": float(rfm.frequency) if rfm else 0,
                "z": float(rfm.monetary) if rfm else 0
            })
        cls_names = ["高消费活跃用户", "中等消费用户", "潜力用户", "低频用户", "流失用户"]
        clusters = []
        for label, pts in sorted(groups.items()):
            cx = sum(p["x"] for p in pts) / len(pts) if pts else 0
            cy = sum(p["y"] for p in pts) / len(pts) if pts else 0
            cz = sum(p["z"] for p in pts) / len(pts) if pts else 0
            clusters.append({"label": label, "name": cls_names[label] if label < len(cls_names) else f"群{label}",
                           "count": len(pts), "center": [round(cx, 1), round(cy, 1), round(cz, 1)], "points": pts})
        result = {"clusters": clusters}
    
    resp = ClusterResultResponse(clusters=[
        ClusterItem(label=c["label"], name=c["name"], count=c["count"],
                    center=c["center"], points=[ClusterPoint(**p) for p in c["points"]])
        for c in result["clusters"]
    ])
    return success_response(data=resp.model_dump())


# ======================== 7.3 触发重新计算 ========================

@router.post("/recalculate", summary="触发重新计算")
async def recalculate_analysis(
    request: RecalculateRequest,
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """触发后台重新计算RFM和/或聚类结果"""
    results = {}
    if request.type in ("rfm", "all"):
        calc_rfm = calculate_rfm(db)
        results["rfm"] = {"segments": calc_rfm["distribution"]["values"]}
    if request.type in ("cluster", "all"):
        calc_cluster = calculate_cluster(db, request.clusters)
        results["cluster"] = {"groups": len(calc_cluster.get("clusters", []))}
    if request.type not in ("rfm", "cluster", "all"):
        return error_response(code=status.HTTP_400_BAD_REQUEST, message="type必须为 rfm/cluster/all")
    
    return success_response(data=results, message="重新计算完成")


# ======================== 7.4 关联规则挖掘 ========================

@router.get("/association", summary="关联规则挖掘")
async def get_association(
    min_support: float = Query(default=0.01, ge=0.001, le=1.0, description="最小支持度"),
    min_confidence: float = Query(default=0.5, ge=0.1, le=1.0, description="最小置信度"),
    limit: int = Query(default=20, ge=1, le=100, description="返回数量"),
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """
    获取商品关联规则（Apriori算法）
    
    基于订单数据中的商品共现关系，挖掘关联规则
    返回规则包含支持度(support)、置信度(confidence)、提升度(lift)
    """
    rules = calculate_association(db, min_support, min_confidence, limit)
    resp = AssociationResult(rules=[
        AssociationRule(**r) for r in rules
    ])
    return success_response(data=resp.model_dump())


# ======================== 7.5 用户画像 ========================

@router.get("/profile/{user_id}", summary="用户画像")
async def get_profile(
    user_id: int,
    db: Session = Depends(get_db),
    current_user: Dict = Depends(get_current_user)
) -> Dict:
    """获取指定用户的画像数据（偏好分类、活跃时段、标签等）"""
    profile = get_user_profile(db, user_id)
    if not profile:
        return error_response(code=status.HTTP_404_NOT_FOUND, message="用户不存在")
    
    resp = UserProfileResponse(
        user_id=profile["user_id"], nickname=profile["nickname"],
        tags=profile["tags"],
        prefer_categories=[PreferCategory(**c) for c in profile["prefer_categories"]],
        active_hours=profile["active_hours"],
        avg_order_value=profile["avg_order_value"],
        total_orders=profile["total_orders"],
        days_since_last_purchase=profile["days_since_last_purchase"]
    )
    return success_response(data=resp.model_dump())
